crates/ra_assists/src/assists/add_explicit_type.rs

```diff
@@ -34,17 +35,29 @@ pub(crate) fn add_explicit_type(ctx: AssistCtx<impl HirDatabase>) -> Option<Assi
     // The binding must have a name
     let name = pat.name()?;
     let name_range = name.syntax().text_range();
-    // Assist should only be applicable if cursor is between 'let' and '='
     let stmt_range = stmt.syntax().text_range();
     let eq_range = stmt.eq_token()?.text_range();
+    let ty_range =
+        stmt.syntax().children_with_tokens().find(|child| child.kind() == T![:]).map(|token| {
+            TextRange::from_to(
+                token.text_range().start(),
+                eq_range.start().checked_sub(1.into()).unwrap_or(eq_range.start()),
+            )
+        });
+    // Assist should only be applicable if cursor is between 'let' and '='
     let let_range = TextRange::from_to(stmt_range.start(), eq_range.start());
     let cursor_in_range = ctx.frange.range.is_subrange(&let_range);
     if !cursor_in_range {
         return None;
     }
     // Assist not applicable if the type has already been specified
-    if stmt.syntax().children_with_tokens().any(|child| child.kind() == T![:]) {
-        return None;
+    // and it has no placeholders
+    if ty_range.is_some() {
+        if stmt.syntax().children_with_tokens().any(|child| child.kind() == T![:]) {
+            if !stmt.syntax().descendants().any(|child| child.kind() == PLACEHOLDER_TYPE) {
```

Comment:
I think we should walk only the type , and not the whole statement here (which includes the initalizing expression). 

it seems like the following should do the trick
```
match stmt.ascribed_type() {
    None => /* no type at all, try to insert one */,
    Some(ty) => /* there's already a type, let's check if it contains `_` */
}
```